src/backchannel_logout.rs

```diff
@@ -0,0 +1,543 @@
+//! This module implements components needed for [back-channel logout]
+//!
+//! [back-channel logout]: <https://openid.net/specs/openid-connect-backchannel-1_0.html>
+
+use std::collections::HashMap;
+
+use chrono::{DateTime, Utc};
+use serde::{de::Error, ser::SerializeMap, Deserialize, Serialize};
+
+use crate::{
+    helpers::{FilteredFlatten, FlattenFilter},
+    jwt::{JsonWebToken, JsonWebTokenJsonPayloadSerde},
+    types::helpers::{deserialize_string_or_vec, serde_utc_seconds},
+    types::SessionIdentifier,
+    AdditionalClaims, Audience, IssuerUrl, JsonWebKeyId, JsonWebKeyType,
+    JweContentEncryptionAlgorithm, JwsSigningAlgorithm, SubjectIdentifier,
+};
+
+/// Back-Channel Logout Token
+///
+/// Parses a JWT as a Logout Token as definied in [section 2.4]
+///
+/// [section 2.4]: <https://openid.net/specs/openid-connect-backchannel-1_0.html#LogoutToken>
+#[derive(Debug, Clone, PartialEq)]
+pub struct LogoutToken<AC, JE, JS, JT>(
+    JsonWebToken<JE, JS, JT, LogoutTokenClaims<AC>, JsonWebTokenJsonPayloadSerde>,
+)
+where
+    AC: AdditionalClaims,
+    JE: JweContentEncryptionAlgorithm<JT>,
+    JS: JwsSigningAlgorithm<JT>,
+    JT: JsonWebKeyType;
+
+impl<AC, JE, JS, JT> LogoutToken<AC, JE, JS, JT>
+where
+    AC: AdditionalClaims,
+    JE: JweContentEncryptionAlgorithm<JT>,
+    JS: JwsSigningAlgorithm<JT>,
+    JT: JsonWebKeyType,
+{
+    // TODO: implement signature verification & friends
+}
+
+/// The Logout Token Claims as defined in [section 2.4] of the [OpenID Connect Back-Channel Logout spec][1]
+///
+/// [1]: <https://openid.net/specs/openid-connect-backchannel-1_0.html>
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub struct LogoutTokenClaims<AC: AdditionalClaims> {
+    /// The issuer of this token
+    iss: IssuerUrl,
+    /// The audience this token is intended for
+    aud: Vec<Audience>,
+    /// Time at which this token was issued
+    #[serde(with = "serde_utc_seconds")]
+    iat: DateTime<Utc>,
+    /// The unique identifier for this token. This can be used to detect
+    /// replay attacks.
+    jti: JsonWebKeyId,
+    #[serde(flatten)]
+    identifier: LogoutIdentifier,
+    events: HashMap<String, serde_json::Value>,
+    additional_claims: FilteredFlatten<Self, AC>,
```

Comment:
I think this should be:
```rust
additional_claims: FilteredFlatten<LogoutIdentifier, AC>,
```

The issue we're trying to solve here is closely related to serde-rs/serde#1346. If a struct contains any fields with `#[serde(flatten)]`, then Serde's derive macros treat the entire struct as a map rather than a "struct" in Serde's [data model](https://serde.rs/data-model.html#types).

The way deserialization works in this case is that all of the known field names (e.g., `iss`, `aud`, etc. in `LogoutTokenClaims`) are deserialized to the corresponding fields in the "outer" struct (`LogoutTokenClaims`), and all of the other key-value pairs from the input are gathered into a vec called `__collect`: https://github.com/serde-rs/serde/blob/fb2fe409c8f7ad6c95e3096e5e9ede865c8cfb49/serde_derive/src/de.rs#L2459-L2464.

Then, the key-value pairs in `__collect` are passed to the deserializers for each of the flattened fields. However, Serde doesn't do anything to ensure mutual exclusivity between flattened fields. This means that even though `LogoutIdentifier` will use `sub` and/or `sid`, those fields will _also_ be passed to `additional_claims`'s deserializer. The `FilteredFlatten` struct's job here is to filter out `sub` and `sid` so that they don't end up duplicated in `additional_claims`.

The [`cargo expand`](https://github.com/dtolnay/cargo-expand) output is really helpful for debugging serde behavior.

For more context, see https://github.com/ramosbugs/openidconnect-rs/commit/02c606445375b5b365ad466c96c19f2016aaaf26, which is where I originally fixed this issue for ID tokens and user info claims.